src/Microsoft.ML.TimeSeries/IidAnomalyDetectionBase.cs

```diff
@@ -13,99 +13,147 @@
+
+        public IDataView Transform(IDataView input) => Base.Transform(input);
 
-        public override Schema GetOutputSchema(Schema inputSchema)
+        public virtual void Save(ModelSaveContext ctx)
         {
-            Host.CheckValue(inputSchema, nameof(inputSchema));
+            Base.SaveThis(ctx);
+        }
 
-            if (!inputSchema.TryGetColumnIndex(InputColumnName, out var col))
-                throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", InputColumnName);
+        internal IStatefulRowMapper MakeRowMapper(Schema schema) => Base.MakeRowMapper(schema);
 
-            var colType = inputSchema[col].Type;
-            if (colType != NumberType.R4)
-                throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", InputColumnName, "float", colType.ToString());
+        internal IDataTransform MakeDataTransform(IDataView input) => Base.MakeDataTransform(input);
 
-            return Transform(new EmptyDataView(Host, inputSchema)).Schema;
+        internal IidAnomalyDetectionBase Base;
+        public IidAnomalyDetectionBaseWrapper(ArgumentsBase args, string name, IHostEnvironment env)
+        {
+            Base = new IidAnomalyDetectionBase(args, name, env, this);
         }
 
-        public override void Save(ModelSaveContext ctx)
+        public IidAnomalyDetectionBaseWrapper(IHostEnvironment env, ModelLoadContext ctx, string name)
         {
-            ctx.CheckAtModel();
-            Host.Assert(InitialWindowSize == 0);
-            base.Save(ctx);
-
-            // *** Binary format ***
-            // <base>
-            // State: StateRef
-            StateRef.Save(ctx.Writer);
+            Base = new IidAnomalyDetectionBase(env, ctx, name, this);
         }
 
-        public sealed class State : AnomalyDetectionStateBase
+        /// <summary>
+        /// This transform computes the p-values and martingale scores for a supposedly i.i.d input sequence of floats. In other words, it assumes
+        /// the input sequence represents the raw anomaly score which might have been computed via another process.
+        /// </summary>
+        internal class IidAnomalyDetectionBase : SequentialAnomalyDetectionTransformBase<Single, IidAnomalyDetectionBase.State>
         {
-            public State()
-            {
-            }
+            internal IidAnomalyDetectionBaseWrapper Parent;
 
-            internal State(BinaryReader reader) : base(reader)
+            public IidAnomalyDetectionBase(ArgumentsBase args, string name, IHostEnvironment env, IidAnomalyDetectionBaseWrapper parent)
+                : base(args, name, env)
             {
-                WindowedBuffer = TimeSeriesUtils.DeserializeFixedSizeQueueSingle(reader, Host);
-                InitialWindowedBuffer = TimeSeriesUtils.DeserializeFixedSizeQueueSingle(reader, Host);
+                InitialWindowSize = 0;
+                StateRef = new State();
```

Comment:
The sementic of these two expressions looks a bit not smooth. Their meaning sounds like we `create` a `State` and then `Initialize` it. Maybe we can combine them by creating a constructor?

Reply:
It more than that. InitiState comes from an interface that state class implements. There are also cases where we don';t call Init. 

---
In reply to: [254480705](https://github.com/dotnet/machinelearning/pull/2344#discussion_r254480705) [](ancestors = 254480705)